In [1]:
import numpy as np
import matplotlib.pyplot as plt
# to plot accuracy and loss diagrams
from tensorflow import keras
# implement keras Neural Network
from keras import layers
import os
# used to find download path
import pandas as pd

In [2]:
file_path = '/content/concrete_data.csv'
df = pd.read_csv(file_path)
df.head()
#get our dataframe from the csv file

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

In [3]:
in_features = [
            'Cement',
            'Blast Furnace Slag',
            'Fly Ash',
            'Water',
            'Superplasticizer',
            'Coarse Aggregate',
            'Fine Aggregate',
            'Age'
]

out_features = [
                'Strength'
]
# get the input features and target labels

In [4]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(df[in_features].to_numpy(), df[out_features].to_numpy(),
                                                    test_size=0.3, shuffle=True)
x_train, x_test = x_train.astype('float32'), x_test.astype('float32')
y_train, y_test = y_train.astype('float32'), y_test.astype('float32')

# I will use scikitlearn standardscaler to perform standard scaling(aka normalization)

In [5]:
from sklearn.preprocessing import StandardScaler


scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)



In [6]:
print(x_train.shape, x_test.shape)
print(x_train[0])
#successfully train test splitted, show sample of one training data

(721, 8) (309, 8)
[-1.2357584 -0.8633568  2.051226  -0.7133786  1.5779438  1.3352689
 -1.2209537 -0.2816736]


In [7]:
Network = keras.Sequential([
                            keras.Input(shape=(8, )),
                            layers.Dense(10, activation="relu", name="layer1"),
                            layers.Dense(10, activation="relu", name="layer2"),
                            layers.Dense(10, activation="relu", name="layer3"),
                            layers.Dense(1, name="output"),
    ])
#create our network

In [8]:
Network.compile(loss='mse', optimizer= keras.optimizers.Adam(), metrics=['mse'])
Network.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 layer1 (Dense)              (None, 10)                90        
                                                                 
 layer2 (Dense)              (None, 10)                110       
                                                                 
 layer3 (Dense)              (None, 10)                110       
                                                                 
 output (Dense)              (None, 1)                 11        
                                                                 
Total params: 321
Trainable params: 321
Non-trainable params: 0
_________________________________________________________________


### Got the model to train it on the data with mse(Mean_Squared_Error) Optimzer is Adam




In [9]:
def Train(Model, model_nums, epochs):
  Loss_array = np.ndarray([])
  score = np.ndarray([])
  for i in range(model_nums):

    model = keras.models.clone_model(Model)
    #refreshing our model every time
    
    model.compile(loss='mse', optimizer= keras.optimizers.Adam(), metrics=['mse'])
    model.fit(x=x_train, y=y_train, epochs=epochs, batch_size=64, verbose=1)

    #getting our prediction
    res = model.evaluate(x=x_test, y=y_test)
    score = np.hstack([score, res[0]])
  return score


In [10]:
%%time
model_num = 50
epochs_num = 50
res = Train(Network, model_num, epochs_num)

12/12 [==============================] - 1s 4ms/step - loss: 1539.5363 - mse: 1539.5363
Epoch 2/50
12/12 [==============================] - 0s 4ms/step - loss: 1529.2190 - mse: 1529.2190
Epoch 3/50
12/12 [==============================] - 0s 3ms/step - loss: 1516.2458 - mse: 1516.2458
Epoch 4/50
12/12 [==============================] - 0s 3ms/step - loss: 1498.4335 - mse: 1498.4335
Epoch 5/50
12/12 [==============================] - 0s 4ms/step - loss: 1475.9370 - mse: 1475.9370
Epoch 6/50
12/12 [==============================] - 0s 4ms/step - loss: 1447.1230 - mse: 1447.1230
Epoch 7/50
12/12 [==============================] - 0s 4ms/step - loss: 1410.9011 - mse: 1410.9011
Epoch 8/50
12/12 [==============================] - 0s 5ms/step - loss: 1362.5721 - mse: 1362.5721
Epoch 9/50
12/12 [==============================] - 0s 4ms/step - loss: 1302.2770 - mse: 1302.2770
Epoch 10/50
12/12 [==============================] - 0s 4ms/step - loss: 1224.7053 - mse: 1224.7053
Epoch 11/50
12/12 [=

There was too much output for the training so i just removed it

In [12]:
print(np.mean(res), '\n', np.std(res))

152.77576371735216 
 23.519211582088186


# These are our results, It worked SPECTACULARLY!!
mean :152.77576371735216 
 
 
standard deviation: 23.519211582088186